### PR TITLE
Nullability and modernization updates, part 3

### PR DESCRIPTION
**Pull request by erikmav**
For #124

### CODE REVIEW

**Comment by erikmav** [comment]
For #124

**Comment by erikmav** [comment]
@FreeAndNil Important to closely review `ReadOnlyPropertiesDictionary` and `PropertiesDictionary`. I selected the most backward compatible option, which is to coerce the `this[]` getter to avoid throwing KeyNotFoundException like Dictionary<> would (including a new test for the behavior). Also note that to avoid a double-indirection it would be ideal to change both to structs.

**Comment by FreeAndNil** [comment]
Thanks. I will review on Monday.

File: src/log4net/Util/ThreadContextStack.cs [hunk 1]
```diff
@@ -223,7 +208,7 @@ internal Stack InternalStack
     /// Gets the current context information for this stack.
     /// </para>
     /// </remarks>
-    public override string ToString()
+    public override string? ToString()
```
Comment (FreeAndNil) [inline src/log4net/Util/ThreadContextStack.cs:211]:
```suggestion
    public override string? ToString() => GetFullMessage();
```

File: src/log4net/Util/ThreadContextStack.cs [hunk 2]
```diff
@@ -239,7 +224,7 @@ public override string ToString()
     /// </remarks>
     object IFixingRequired.GetFixedObject()
```
Comment (FreeAndNil) [inline src/log4net/Util/ThreadContextStack.cs:225]:
```suggestion
    object IFixingRequired.GetFixedObject() => GetFullMessage() ?? string.Empty;
```

File: src/log4net/Util/TypeConverters/ConverterRegistry.cs [hunk 3]
```diff
@@ -208,24 +180,19 @@ public static IConvertFrom GetConvertFrom(Type destinationType)
...
       foreach (var attribute in attributes)
       {
-        var tcAttr = attribute as TypeConverterAttribute;
-        if (tcAttr != null)
+        if (attribute is TypeConverterAttribute tcAttr)
         {
-          var converterType = SystemInfo.GetTypeFromString(destinationType, tcAttr.ConverterTypeName, false, true);
-          return CreateConverterInstance(converterType);
+          Type? converterType = SystemInfo.GetTypeFromString(destinationType, tcAttr.ConverterTypeName, false, true);
+          if (converterType is not null)
```
Comment (FreeAndNil) [inline src/log4net/Util/TypeConverters/ConverterRegistry.cs:192]:
```suggestion
          if (SystemInfo.GetTypeFromString(destinationType, tcAttr.ConverterTypeName, false, true) is Type converterType)
```

**Review by FreeAndNil — CHANGES_REQUESTED** [review]
Thanks. I've made a few suggestions.

File: src/log4net/Core/LoggingEvent.cs [hunk 4]
```diff
@@ -826,11 +825,10 @@ private static string TryReadWindowsIdentityUserName()
         {
           try
           {
-            if (System.Threading.Thread.CurrentPrincipal is not null &&
-                System.Threading.Thread.CurrentPrincipal.Identity is not null &&
-                System.Threading.Thread.CurrentPrincipal.Identity.Name is not null)
+            if (Thread.CurrentPrincipal is not null &&
```
Comment (FreeAndNil) [inline src/log4net/Core/LoggingEvent.cs:828]: I think we can shorten this even more:
```suggestion
            if (Thread.CurrentPrincipal?.Identity?.Name is string name)
```

**Review by FreeAndNil — CHANGES_REQUESTED** [review]
@erikmav one tiny thing left.

**Comment by FreeAndNil** [comment]
> Also note that to avoid a double-indirection it would be ideal to change both to structs.
@erikmav can you explain how you would implement this? Maybe as separate issue?

**Comment by erikmav** [comment]
Could be done under another issue. The change is to change from class to struct (or readonly struct, depending) for ReadOnlyPropertiesDictionary, then derive PropertiesDictionary from that as a second struct. I wouldn't do it in this PR either way, it's mainly an optimization.

**Review by FreeAndNil — APPROVED** [review]
Thanks a lot.

**Pull request merged**